tests/components/songpal/test_media_player.py

```diff
@@ -0,0 +1,267 @@
+    with _patch_media_player_device(mocked_device):
+        await async_setup_component(
+            hass,
+            media_player.DOMAIN,
+            {
+                media_player.DOMAIN: [
+                    {
+                        "platform": songpal.DOMAIN,
+                        CONF_NAME: FRIENDLY_NAME,
+                        CONF_ENDPOINT: ENDPOINT,
+                    }
+                ],
+            },
+        )
+        await hass.async_block_till_done()
+
+    # No device is set up
+    mocked_device.assert_not_called()
+    all_states = hass.states.async_all()
+    assert len(all_states) == 0
+
+
+async def test_setup_failed(hass, caplog):
+    """Test failed to set up the entity."""
+    mocked_device = _create_mocked_device(throw_exception=True)
+    entry = MockConfigEntry(domain=songpal.DOMAIN, data=CONF_DATA)
+    entry.add_to_hass(hass)
+
+    with _patch_media_player_device(mocked_device):
+        await hass.config_entries.async_setup(entry.entry_id)
+        await hass.async_block_till_done()
+    all_states = hass.states.async_all()
+    assert len(all_states) == 0
+    warning_records = [x for x in caplog.records if x.levelno == logging.WARNING]
+    assert len(warning_records) == 2
+    assert not any(x.levelno == logging.ERROR for x in caplog.records)
+    caplog.clear()
+
+    utcnow = dt_util.utcnow()
+    type(mocked_device).get_supported_methods = AsyncMock()
+    with _patch_media_player_device(mocked_device):
+        async_fire_time_changed(hass, utcnow + timedelta(seconds=30))
+        await hass.async_block_till_done()
+    all_states = hass.states.async_all()
+    assert len(all_states) == 1
+    assert not any(x.levelno == logging.WARNING for x in caplog.records)
+    assert not any(x.levelno == logging.ERROR for x in caplog.records)
+
+
+async def test_state(hass):
+    """Test state of the entity."""
+    mocked_device = _create_mocked_device()
+    entry = MockConfigEntry(domain=songpal.DOMAIN, data=CONF_DATA)
+    entry.add_to_hass(hass)
+
+    with _patch_media_player_device(mocked_device):
+        await hass.config_entries.async_setup(entry.entry_id)
+        await hass.async_block_till_done()
+
+    state = hass.states.get(ENTITY_ID)
+    assert state.name == FRIENDLY_NAME
+    assert state.state == STATE_ON
+    attributes = state.as_dict()["attributes"]
+    assert attributes["volume_level"] == 0.5
+    assert attributes["is_volume_muted"] is False
+    assert attributes["source_list"] == ["title1", "title2"]
+    assert attributes["source"] == "title2"
+    assert attributes["supported_features"] == SUPPORT_SONGPAL
+
+    device_registry = await dr.async_get_registry(hass)
+    device = device_registry.async_get_device(
+        identifiers={(songpal.DOMAIN, MAC)}, connections={}
+    )
+    assert device.connections == {(dr.CONNECTION_NETWORK_MAC, MAC)}
+    assert device.manufacturer == "Sony Corporation"
+    assert device.name == FRIENDLY_NAME
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5518609)